sqlmesh/core/snapshot/definition.py

```diff
@@ -1040,12 +1011,166 @@ class SnapshotTableCleanupTask(PydanticModel):
 SnapshotNameVersionLike = t.Union[SnapshotNameVersion, SnapshotTableInfo, Snapshot]
 
 
-def table_name(physical_schema: str, name: str, version: str, is_temp: bool = False) -> str:
+class DeployabilityIndex(PydanticModel, frozen=True):
+    """Contains information about deployability of every snapshot.
+
+    Deployability is defined as whether or not the output that a snapshot produces during the
+    current evaluation can be reused in (deployed to) the production environment.
+    """
+
+    indexed_ids: t.FrozenSet[str]
+    is_opposite_index: bool = False
+    representative_shared_version_ids: t.Optional[t.FrozenSet[str]] = None
+
+    @field_validator("indexed_ids", "representative_shared_version_ids", mode="before")
+    @classmethod
+    def _snapshot_ids_set_validator(cls, v: t.Any) -> t.Optional[t.FrozenSet[t.Tuple[str, str]]]:
+        if v is None:
+            return v
+        # Transforming into strings because the serialization of sets of objects / lists is broken in Pydantic.
+        return frozenset(
+            {
+                cls._snapshot_id_key(snapshot_id)
+                if isinstance(snapshot_id, SnapshotId)
+                else snapshot_id
+                for snapshot_id in v
+            }
+        )
+
+    def is_deployable(self, snapshot: SnapshotIdLike) -> bool:
+        """Returns true if the output produced by the given snapshot in a development environment can be reused
+        in (deployed to) production
+
+        Args:
+            snapshot: The snapshot to check.
+
+        Returns:
+            True if the snapshot is deployable, False otherwise.
+        """
+        snapshot_id = self._snapshot_id_key(snapshot.snapshot_id)
```

Comment:
return (self.is_opposite_index and snapshot_id not in self.indexed_ids) or (not self.is_opposite and snapshot_id in self.indexed_ids)